5.15内核，编译错误

内核版本5.15，gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04) 


KernelSU module (KSU) [Y/n/m/?] (NEW) y
  KernelSU module debug mode (KSU_DEBUG) [N/m/y/?] (NEW) n
  DESCEND objtool
  CALL    scripts/atomic/check-atomics.sh
  CALL    scripts/checksyscalls.sh
  CHK     include/generated/compile.h
  UPD     kernel/config_data
  GZIP    kernel/config_data.gz
  CC      kernel/configs.o
  AR      kernel/built-in.a
  CC      drivers/kernelsu/ksu.o
  CC      drivers/kernelsu/selinux/selinux.o
  CC      drivers/kernelsu/selinux/sepolicy.o
In file included from drivers/kernelsu/ksu.c:9:0:
drivers/kernelsu/ksu.h:7:2: warning: #warning "KSU_GIT_VERSION not defined! It is better to make KernelSU a git submodule!" [-Wcpp]
 #warning "KSU_GIT_VERSION not defined! It is better to make KernelSU a git submodule!"
  ^~~~~~~
cc1: warning: unrecognized command line option ‘-Wno-macro-redefined’
cc1: warning: unrecognized command line option ‘-Wno-gcc-compat’
  CC      drivers/kernelsu/allowlist.o
In file included from ./include/linux/kernel.h:19:0,
                 from ./include/linux/delay.h:22,
                 from drivers/kernelsu/allowlist.c:1:
drivers/kernelsu/allowlist.c: In function ‘do_persistent_allow_list’:
./include/linux/kern_levels.h:5:18: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
 #define KERN_SOH "\001"  /* ASCII Start Of Header */
                  ^
./include/linux/printk.h:421:11: note: in definition of macro ‘printk_index_wrap’
   _p_func(_fmt, ##__VA_ARGS__);    \
           ^~~~
./include/linux/printk.h:492:2: note: in expansion of macro ‘printk’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
  ^~~~~~
./include/linux/kern_levels.h:11:18: note: in expansion of macro ‘KERN_SOH’
 #define KERN_ERR KERN_SOH "3" /* error conditions */
                  ^~~~~~~~
./include/linux/printk.h:492:9: note: in expansion of macro ‘KERN_ERR’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
         ^~~~~~~~
drivers/kernelsu/allowlist.c:127:3: note: in expansion of macro ‘pr_err’
   pr_err("save_allow_list creat file failed: %d\n", PTR_ERR(fp));
   ^~~~~~
drivers/kernelsu/allowlist.c:127:47: note: format string is defined here
   pr_err("save_allow_list creat file failed: %d\n", PTR_ERR(fp));
                                              ~^
                                              %ld
In file included from ./include/linux/kernel.h:19:0,
                 from ./include/linux/delay.h:22,
                 from drivers/kernelsu/allowlist.c:1:
drivers/kernelsu/allowlist.c: In function ‘do_load_allow_list’:
./include/linux/kern_levels.h:5:18: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
 #define KERN_SOH "\001"  /* ASCII Start Of Header */
                  ^
./include/linux/printk.h:421:11: note: in definition of macro ‘printk_index_wrap’
   _p_func(_fmt, ##__VA_ARGS__);    \
           ^~~~
./include/linux/printk.h:492:2: note: in expansion of macro ‘printk’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
  ^~~~~~
./include/linux/kern_levels.h:11:18: note: in expansion of macro ‘KERN_SOH’
 #define KERN_ERR KERN_SOH "3" /* error conditions */
                  ^~~~~~~~
./include/linux/printk.h:492:9: note: in expansion of macro ‘KERN_ERR’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
         ^~~~~~~~
drivers/kernelsu/allowlist.c:166:3: note: in expansion of macro ‘pr_err’
   pr_err("load_allow_list open '/data/adb': %d\n", PTR_ERR(fp));
   ^~~~~~
drivers/kernelsu/allowlist.c:166:46: note: format string is defined here
   pr_err("load_allow_list open '/data/adb': %d\n", PTR_ERR(fp));
                                             ~^
                                             %ld
In file included from ./include/linux/kernel.h:19:0,
                 from ./include/linux/delay.h:22,
                 from drivers/kernelsu/allowlist.c:1:
./include/linux/kern_levels.h:5:18: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
 #define KERN_SOH "\001"  /* ASCII Start Of Header */
                  ^
./include/linux/printk.h:421:11: note: in definition of macro ‘printk_index_wrap’
   _p_func(_fmt, ##__VA_ARGS__);    \
           ^~~~
./include/linux/printk.h:492:2: note: in expansion of macro ‘printk’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
  ^~~~~~
./include/linux/kern_levels.h:11:18: note: in expansion of macro ‘KERN_SOH’
 #define KERN_ERR KERN_SOH "3" /* error conditions */
                  ^~~~~~~~
./include/linux/printk.h:492:9: note: in expansion of macro ‘KERN_ERR’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
         ^~~~~~~~
drivers/kernelsu/allowlist.c:192:3: note: in expansion of macro ‘pr_err’
   pr_err("load_allow_list open file failed: %d\n", PTR_ERR(fp));
   ^~~~~~
drivers/kernelsu/allowlist.c:192:46: note: format string is defined here
   pr_err("load_allow_list open file failed: %d\n", PTR_ERR(fp));
                                             ~^
                                             %ld
In file included from ./include/linux/kernel.h:19:0,
                 from ./include/linux/delay.h:22,
                 from drivers/kernelsu/allowlist.c:1:
./include/linux/kern_levels.h:5:18: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘ssize_t {aka long int}’ [-Wformat=]
 #define KERN_SOH "\001"  /* ASCII Start Of Header */
                  ^
./include/linux/printk.h:421:11: note: in definition of macro ‘printk_index_wrap’
   _p_func(_fmt, ##__VA_ARGS__);    \
           ^~~~
./include/linux/printk.h:522:2: note: in expansion of macro ‘printk’
  printk(KERN_INFO pr_fmt(fmt), ##__VA_ARGS__)
  ^~~~~~
./include/linux/kern_levels.h:14:19: note: in expansion of macro ‘KERN_SOH’
 #define KERN_INFO KERN_SOH "6" /* informational */
                   ^~~~~~~~
./include/linux/printk.h:522:9: note: in expansion of macro ‘KERN_INFO’
  printk(KERN_INFO pr_fmt(fmt), ##__VA_ARGS__)
         ^~~~~~~~~
drivers/kernelsu/allowlist.c:217:4: note: in expansion of macro ‘pr_info’
    pr_info("load_allow_list read err: %d\n", ret);
    ^~~~~~~
drivers/kernelsu/allowlist.c:217:40: note: format string is defined here
    pr_info("load_allow_list read err: %d\n", ret);
                                       ~^
                                       %ld
drivers/kernelsu/allowlist.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-macro-redefined’
cc1: warning: unrecognized command line option ‘-Wno-gcc-compat’
drivers/kernelsu/selinux/sepolicy.c:67:0: warning: "hash_for_each" redefined
 #define hash_for_each(node_ptr, n_slot, cur)                                   \
 
In file included from ./security/selinux/ss/sidtab.h:16:0,
                 from ./security/selinux/ss/policydb.h:27,
                 from drivers/kernelsu/selinux/sepolicy.h:6,
                 from drivers/kernelsu/selinux/sepolicy.c:1:
./include/linux/hashtable.h:126:0: note: this is the location of the previous definition
 #define hash_for_each(name, bkt, obj, member)    \
 
  CC      drivers/kernelsu/apk_sign.o
  CC      drivers/kernelsu/module_api.o
cc1: warning: unrecognized command line option ‘-Wno-macro-redefined’
  CC      drivers/kernelsu/selinux/rules.o
drivers/kernelsu/apk_sign.c: In function ‘check_v2_signature’:
drivers/kernelsu/apk_sign.c:29:2: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
  for (int i = 0;; ++i) {
  ^~~
drivers/kernelsu/apk_sign.c:29:2: note: use option -std=c99, -std=gnu99, -std=c11 or -std=gnu11 to compile your code
drivers/kernelsu/apk_sign.c:105:5: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
     for (unsigned i = 0; i < size4; ++i) {
     ^~~
drivers/kernelsu/apk_sign.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-macro-redefined’
cc1: warning: unrecognized command line option ‘-Wno-gcc-compat’
scripts/Makefile.build:287: recipe for target 'drivers/kernelsu/apk_sign.o' failed
make[2]: *** [drivers/kernelsu/apk_sign.o] Error 1
make[2]: *** 正在等待未完成的任务....
In file included from ./include/linux/kernel.h:19:0,
                 from ./arch/x86/include/asm/percpu.h:27,
                 from ./arch/x86/include/asm/current.h:6,
                 from ./include/linux/sched.h:12,
                 from ./include/linux/uaccess.h:8,
                 from drivers/kernelsu/selinux/rules.c:1:
drivers/kernelsu/selinux/rules.c: In function ‘handle_sepolicy’:
./include/linux/kern_levels.h:5:18: warning: format ‘%d’ expects a matching ‘int’ argument [-Wformat=]
 #define KERN_SOH "\001"  /* ASCII Start Of Header */
                  ^
./include/linux/printk.h:421:11: note: in definition of macro ‘printk_index_wrap’
   _p_func(_fmt, ##__VA_ARGS__);    \
           ^~~~
./include/linux/printk.h:492:2: note: in expansion of macro ‘printk’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
  ^~~~~~
./include/linux/kern_levels.h:11:18: note: in expansion of macro ‘KERN_SOH’
 #define KERN_ERR KERN_SOH "3" /* error conditions */
                  ^~~~~~~~
./include/linux/printk.h:492:9: note: in expansion of macro ‘KERN_ERR’
  printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
         ^~~~~~~~
drivers/kernelsu/selinux/rules.c:454:3: note: in expansion of macro ‘pr_err’
   pr_err("sepol: unknown cmd: %d");
   ^~~~~~
drivers/kernelsu/selinux/rules.c:454:32: note: format string is defined here
   pr_err("sepol: unknown cmd: %d");
                               ~^
drivers/kernelsu/selinux/rules.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-macro-redefined’
  AR      drivers/kernelsu/selinux/built-in.a
scripts/Makefile.build:549: recipe for target 'drivers/kernelsu' failed
make[1]: *** [drivers/kernelsu] Error 2
make[1]: *** 正在等待未完成的任务....
Makefile:1956: recipe for target 'drivers' failed
make: *** [drivers] Error 2

drivers/kernelsu/apk_sign.c:29:2: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
for (int i = 0;; ++i) {
^~~
drivers/kernelsu/apk_sign.c:29:2: note: use option -std=c99, -std=gnu99, -std=c11 or -std=gnu11 to compile your code
drivers/kernelsu/apk_sign.c:105:5: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
for (unsigned i = 0; i < size4; ++i) {
^~~

--- a/kernel/apk_sign.c
+++ b/kernel/apk_sign.c
@@ -93,7 +93,7 @@ check_v2_signature(char *path, unsigned expected_size, unsigned expected_hash)
 #if 0
 			int hash = 1;
 			signed char c;
-			for (unsigned i = 0; i < size4; ++i) {
+			for (i = 0; i < size4; ++i) {
 				ksu_kernel_read_compat(fp, &c, 0x1, &pos);
 				hash = 31 * hash + c;
 			}
@@ -103,7 +103,7 @@ check_v2_signature(char *path, unsigned expected_size, unsigned expected_hash)
 			if (size4 == expected_size) {
 				int hash = 1;
 				signed char c;
-				for (unsigned i = 0; i < size4; ++i) {
+				for (i = 0; i < size4; ++i) {
 					ksu_kernel_read_compat(fp, &c, 0x1, &pos);
 					hash = 31 * hash + c;
 				}

